### PR TITLE
Update state access

### PR DESCRIPTION
**Pull request by lunfardo314**
refactors state updates, blocks and virtual state. 
Simplifies state manager logic and optimizes it

### CODE REVIEW

File: go.sum [hunk 1]
```diff
@@ -433,10 +436,14 @@ github.com/iotaledger/goshimmer v0.5.6-0.20210423101223-bd2896e40a80 h1:pqNbi/k+
 github.com/iotaledger/goshimmer v0.5.6-0.20210423101223-bd2896e40a80/go.mod h1:qB5ys8qq1LwmqKxfBX/2oE61ebiQKUX2yU3RDGrR2Dw=
 github.com/iotaledger/goshimmer v0.5.6-0.20210424103137-4ed3e27a9622 h1:iTZzYVp4ZKj9GCn6Ws/IYTfg3vJJDVp0i3zxcJgkMjg=
 github.com/iotaledger/goshimmer v0.5.6-0.20210424103137-4ed3e27a9622/go.mod h1:qB5ys8qq1LwmqKxfBX/2oE61ebiQKUX2yU3RDGrR2Dw=
+github.com/iotaledger/goshimmer v0.5.7-0.20210502183158-8fffbc9e00f2 h1:RisMG8d9MfsbEumNdygeGAmBeJigZma+dkTnt0qINII=
+github.com/iotaledger/goshimmer v0.5.7-0.20210502183158-8fffbc9e00f2/go.mod h1:ZYdi8Pmd9KiPOoqGY2pxYor3JibPCZJ2zyoW2orUd3M=
```
Comment (dessaya) [inline go.sum]: Tip: run `go mod tidy` to optimize `go.sum` (e.g. remove old goshimmer versions)

File: packages/kv/codec/encodego.go [hunk 2]
```diff
@@ -3,6 +3,7 @@ package codec
 import (
 	"fmt"
 	"github.com/iotaledger/goshimmer/packages/ledgerstate"
+	"time"
 
 	"github.com/iotaledger/wasp/packages/coretypes"
 	"github.com/iotaledger/wasp/packages/hashing"
```
Comment (dessaya) [inline packages/kv/codec/encodego.go]: Tip: configure your IDE to run `goimports` so that imports are sorted and grouped according to https://github.com/golang/go/wiki/CodeReviewComments#imports

File: packages/chain/statemgr/action.go [hunk 3]
```diff
@@ -17,113 +16,114 @@ func (sm *stateManager) takeAction() {
...
-func (sm *stateManager) checkStateApproval() {
+func (sm *stateManager) checkStateTransition() {
+	if sm.isSynced() {
+		return
+	}
 	if sm.stateOutput == nil {
 		return
 	}
-	// among candidate state update batches we locate the one which
+	// among stateCandidate state update batches we locate the one which
```
Comment (dessaya) [inline packages/chain/statemgr/action.go]: was this change intentional?

File: packages/kv/codec/time.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+	return time.Unix(0, nanos), true, nil
+}
+
+var b8 [8]byte
+
+func EncodeTime(value time.Time) []byte {
+	nanos := value.UnixNano()
+	if nanos == zeroUnixNano {
+		return b8[:]
+	}
```
Comment (dessaya) [inline packages/kv/codec/time.go:38]: Why the special case?

File: packages/kv/codec/time.go [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+		panic("inconsistency: zeroUnixNano != -6795364578871345152")
+	}
+}
+
+func DecodeTime(b []byte) (time.Time, bool, error) {
+	if b == nil {
+		// special behavior for backward compatibility: nil value is treated as the absence of a value, not an error
+		return time.Time{}, false, nil
+	}
+	nanos, err := util.Int64From8Bytes(b)
```
Comment (dessaya) [inline packages/kv/codec/time.go:22]: `time.Time` already has `MarshalBinary()` and `UnmarshalBinary()` methods. Use those instead?

File: packages/vm/viewcontext/sandbox.go [hunk 6]
```diff
@@ -86,7 +86,7 @@ func (s *sandboxview) ContractCreator() *coretypes.AgentID {
 }
 
 func (s *sandboxview) GetTimestamp() int64 {
```
Comment (dessaya) [inline packages/vm/viewcontext/sandbox.go:88]: Shouldn't this just return `time.Time`? (maybe for a future PR)

File: tools/wasp-cli/dashboard/base.go [hunk 7]
```diff
@@ -109,7 +109,7 @@ const TplSCInfo = `
     <p>
 	  <ul>
 		<li>Index: <code>{{.Status.BlockIndex}}</code></li>
-		<li>Timestamp: <code>{{.Status.Timestamp}}</code></li>
+		<li>OutputTimestamp: <code>{{.Status.OutputTimestamp}}</code></li>
```
Comment (dessaya) [inline tools/wasp-cli/dashboard/base.go]: Is this change intended? (this file is build ignored)

**Review by dessaya — APPROVED** [review]

File: packages/kv/codec/encodego.go (hunk 2 above)
Comment (lunfardo314) [reply packages/kv/codec/encodego.go]: I only found how it can run on a particular file

File: packages/kv/codec/time.go (hunk 4 above)
Comment (lunfardo314) [reply packages/kv/codec/time.go:38]: it is a constant. I took it from goshimer code :P

File: packages/kv/codec/time.go (hunk 5 above)
Comment (lunfardo314) [reply packages/kv/codec/time.go:22]: made it compatible with Goshimmer. But I will check

File: packages/vm/viewcontext/sandbox.go (hunk 6 above)
Comment (lunfardo314) [reply packages/vm/viewcontext/sandbox.go:88]: agree, not everything refactored to Goshimmer style yet

File: packages/kv/codec/time.go (hunk 4 above)
Comment (lunfardo314) [reply packages/kv/codec/time.go:38]: in order to return all 0
